.github/workflows/test-go-task.yml

```
@@ -92,11 +92,25 @@ jobs:
           GO_MODULE_PATH: ${{ matrix.module.path }}
         run: task go:test
 
+      # A token is used to avoid intermittent spurious job failures caused by rate limiting.
+      - name: Set up Codecov upload token
+        run: |
```

```suggestion
        shell: bash
        run: |
```

This will cause the Bash shell to be used on the Windows runner in addition to the Linux and macOS runners, so that the Bash shell syntax can be used cross-platform.

---

> What about using[ conditional statement](https://docs.github.com/en/actions/using-jobs/using-conditions-to-control-job-execution) inside github actions?

Are you referring to replacing the Bash conditional with a GitHub Actions conditional? The reason I didn't take that approach in [the code this PR was based on](https://github.com/arduino/arduino-cli/pull/2129#pullrequestreview-1362270583) is because I felt it was more difficult to understand or maintain.

---

With the most minimal version of the GitHub Actions conditional-based approach, it is not clear what the `token` input will be set to when the "Set up Codecov upload token" job did not run. It relies on the somewhat non-obvious and obscure GitHub Actions behavior of [returning an empty string when a non-existent property is referenced](https://docs.github.com/en/actions/learn-github-actions/contexts#:~:text=If%20you%20attempt%20to%20dereference%20a%20non%2Dexistent%20property%2C%20it%20will%20evaluate%20to%20an%20empty%20string.)


```yaml
# A token is used to avoid intermittent spurious job failures caused by rate limiting.
- name: Set up Codecov upload token
  if: github.repository == 'arduino/pluggable-discovery-protocol-handler'
  run: |
    # In order to avoid uploads of data from forks, only use the token for runs in the arduino/arduino-ide repo.
    # Token is intentionally exposed.
    # See: https://community.codecov.com/t/upload-issues-unable-to-locate-build-via-github-actions-api/3954
    echo "CODECOV_TOKEN=d2497f03-291d-4e0c-9d31-d94614ed7c16" >> "$GITHUB_ENV"

- name: Send unit tests coverage to Codecov
  if: runner.os == 'Linux'
  uses: codecov/codecov-action@v3
  with:
    token: ${{ env.CODECOV_TOKEN }}
```

---

There are some alternatives that do not rely on the non-existent property referencing behavior. They may be easier to understand than the minimal version, but I feel they end up being more complex than the shell-based approach in the end:

```yaml
# A token is used to avoid intermittent spurious job failures caused by rate limiting.
- name: Set up Codecov upload token for runs within the repository
  if: github.repository == 'arduino/pluggable-discovery-protocol-handler'
  run: |
    # In order to avoid uploads of data from forks, only use the token for runs in the arduino/arduino-ide repo.
    # Token is intentionally exposed.
    # See: https://community.codecov.com/t/upload-issues-unable-to-locate-build-via-github-actions-api/3954
    echo "CODECOV_TOKEN=d2497f03-291d-4e0c-9d31-d94614ed7c16" >> "$GITHUB_ENV"

- name: Set up Codecov upload token for runs in forks
  if: github.repository != 'arduino/arduino-cli'
  run: |
    # codecov/codecov-action does unauthenticated upload if empty string is passed via the `token` input.
    echo "CODECOV_TOKEN=" >> "$GITHUB_ENV"

- name: Send unit tests coverage to Codecov
  if: runner.os == 'Linux'
  uses: codecov/codecov-action@v3
  with:
    token: ${{ env.CODECOV_TOKEN }}
```

or perhaps this (untested):

```yaml
  test:
    name: test (${{ matrix.module.path }} - ${{ matrix.operating-system }})
    needs: run-determination
    if: needs.run-determination.outputs.result == 'true'

    env:
      # Set default Codecov upload token value. This will be overridden when the job runs within the repository.
      # codecov/codecov-action does unauthenticated upload if empty string is passed via the `token` input.
      CODECOV_TOKEN: ""

    [...]

    steps:

      [...]

      # A token is used to avoid intermittent spurious job failures caused by rate limiting.
      - name: Set up Codecov upload token for runs within the repository
        if: github.repository == 'arduino/pluggable-discovery-protocol-handler'
        run: |
          # In order to avoid uploads of data from forks, only use the token for runs in the arduino/arduino-ide repo.
          # Token is intentionally exposed.
          # See: https://community.codecov.com/t/upload-issues-unable-to-locate-build-via-github-actions-api/3954
          echo "CODECOV_TOKEN=d2497f03-291d-4e0c-9d31-d94614ed7c16" >> "$GITHUB_ENV"

      - name: Send unit tests coverage to Codecov
        if: runner.os == 'Linux'
        uses: codecov/codecov-action@v3
        with:
          token: ${{ env.CODECOV_TOKEN }}
```

---

I'd remove the env variable and use directly the condition:

```yaml
- name: Send unit tests coverage to Codecov
  if: runner.os == 'Linux' && github.repository == 'arduino/pluggable-discovery-protocol-handler'
  uses: codecov/codecov-action@v3
  with:
    # A token is used to avoid intermittent spurious job failures caused by rate limiting.
    token: d2497f03-291d-4e0c-9d31-d94614ed7c16

- name: Send unit tests coverage to Codecov (run from a fork)
  if: runner.os == 'Linux' && github.repository != 'arduino/pluggable-discovery-protocol-handler'
  uses: codecov/codecov-action@v3
  with:
    # use repository variable CODECOV_TOKEN if running from a fork.
    # codecov/codecov-action does unauthenticated upload if empty string is passed via the `token` input.
    token: ${{ var.CODECOV_TOKEN }}
```